src/DataStructures/DataBox/DataBoxTag.hpp

```diff
@@ -120,159 +108,20 @@ struct ConvertToConst<std::unique_ptr<T>> {
   using type = const T&;
 };
 
-template <typename T>
-const T& convert_to_const_type(const T& item) noexcept {
-  return item;
-}
-
-template <typename T>
-const T& convert_to_const_type(const std::unique_ptr<T>& item) noexcept {
-  return *item;
-}
-
-template <typename TagList, typename Tag>
-struct storage_type_impl;
-
-template <int SelectTagType>
-struct dispatch_storage_type;
-
-template <typename ArgsList>
-struct compute_item_type;
-
-template <typename TagList, typename Tag>
-struct storage_type_impl {
-  // storage_type_impl is intentionally a lazy metafunction rather than a
-  // metaclosure or a metacontinuation. The reason is that it is quite likely we
-  // call `db::item_type<Tag>` multiple times within the same translation unit
-  // and so we want to memoize the resulting type. However, we do not want to
-  // memoize the dispatch calls.
-  using type = typename dispatch_storage_type<
-      is_base_tag_v<Tag>
-          ? 4
-          : (is_compute_tag_v<Tag>and has_return_type_member_v<Tag>)
-                ? 3
-                : is_immutable_item_tag_v<Tag>
-                      ? 2
-                      : std::is_base_of_v<db::SimpleTag, Tag> ? 1 : 0>::
-      template f<TagList, Tag>;
-};
-
-template <>
-struct dispatch_storage_type<0> {
-  // Tag is not a tag. This is necessary for SFINAE friendliness, specifically
-  // if someone calls Requires<tt::is_a_v<std::vector, db::item_type<Tag>>>
-  // with, say Tag = double, then this should probably SFINAE away, not fail to
-  // compile.
-  template <typename TagList, typename Tag>
-  using f = NoSuchType;
-};
-
-template <>
-struct dispatch_storage_type<1> {
-  // simple item
-  template <typename TagList, typename Tag>
-  using f = typename Tag::type;
+template <typename Tag, typename TagsList, bool = db::is_base_tag_v<Tag>>
```

Comment:
I'm not sure this is worth doing, but this could be changed to reduce number of types created by doing:
```cpp
template <bool IsBaseTag>
struct const_item_type_impl {
  template <typename Tag, typename TagsList>
  using f = typename db::detail::ConvertToConst<
      std::decay_t<typename Tag::type>>::type;
};

template <>
struct const_item_type_impl<false> {
  template <typename Tag, typename TagsList>
  using f = typename db::detail::ConvertToConst<std::decay_t<
      typename db::detail::first_matching_tag<TagsList, Tag>::type>>::type;
};
```

Reply:
I chose not to do this